internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,217 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+type LogManager interface {
+	PlanChanged(*plan.Plan)
+	NewService(serviceName string, buffer *servicelog.RingBuffer)
+	StopService(serviceName string)
+}
+
+type logManager struct {
+	// Store services and selected log targets, as defined by plan
+	planTargets serviceTargetSet
+	// Store buffers for each running service
+	buffers map[string]*servicelog.RingBuffer
+	// Store log forwarders for each service & target
+	forwarders map[serviceTarget]logForwarder
+	// Log forwarders to be added once the service starts
+	toAdd serviceTargetSet
+}
+
+func NewLogManager() LogManager {
+	return &logManager{
+		buffers:    map[string]*servicelog.RingBuffer{},
+		forwarders: map[serviceTarget]logForwarder{},
+		toAdd:      serviceTargetSet{},
+	}
+}
+
+func (m *logManager) PlanChanged(p *plan.Plan) {
+	m.selectTargets(p)
+	add, remove := deltas(m.planTargets, m.forwarders)
+
+	// Remove obsolete forwarders
+	for st := range remove {
+		m.forwarders[st].Stop()
+		delete(m.forwarders, st)
+	}
+
+	// Add new forwarders
+	m.toAdd.AddAll(add)
+	m.addNewForwarders()
+}
+
+func (m *logManager) selectTargets(p *plan.Plan) {
+	m.planTargets = serviceTargetSet{}
+
+	for serviceName, service := range p.Services {
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if !ok {
+				logger.Noticef("Log manager cannot find target %q", targetName)
+				continue
+			}
+			if target.Selection != plan.DisabledSelection {
+				m.planTargets.Add(serviceTarget{
+					serviceName: serviceName,
+					targetName:  targetName,
+					target:      *target,
+				})
+			}
+		}
+
+		// Add opt-out targets if none specified
+		if len(service.LogTargets) == 0 {
+			for targetName, target := range p.LogTargets {
+				if target.Selection == plan.OptOutSelection {
+					m.planTargets.Add(serviceTarget{
+						serviceName: serviceName,
+						targetName:  targetName,
+						target:      *target,
+					})
+				}
+			}
+		}
+	}
+}
+
+// Register the service (and buffer) with the log manager, and create new
+// forwarders that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
```

Comment:
NewService isn't quite the right name here (as it doesn't return anything, its a notification). Something like `ForwardLogs(serviceName string, buffer *servicelog.RingBuffer, stop <-chan struct{})`, then you can also remove the StopService explicit call.

Reply:
This is the method called by the service manager once the service has started. @benhoyt suggested ServiceStarted as a better name, does that work for you @hpidcock ?

Reply:
Yeah, I found `ForwardLogs` confusing because it sounds like a function that's going to actually do some log forwarding while you wait ... whereas it's actually "notify me that the manager has started a service", hence `ServiceStarted`. Note that it's similar in naming scheme to our other "notify" function, `PlanChanged` -- it tells you what event occurred.